Cython/Compiler/ModuleNode.py

```diff
@@ -54,6 +54,75 @@ def generate_c_code_config(env, options):
         c_line_in_traceback=options.c_line_in_traceback)
 
 
+class IncludeCode(object):
+    """
+    An include file and/or verbatim C code to be included in the
+    generated sources.
+    """
+    # attributes:
+    #
+    #  text      unicode: C code to be generated
+    #  location  int: where to put this include in the C sources, one
+    #            of the constants INITIAL, EARLY, LATE
+    #  order     int: sorting order (automatically set)
+
+    # Constants for location. If the same include occurs with different
+    # locations, the earliest one takes precedense.
+    INITIAL = 0
+    EARLY = 1
+    LATE = 2
+
+    number = 1  # Counter for "order"
+
+    def __new__(cls, *args, **kwds):
+        self = object.__new__(cls)
+        self.order = cls.number
+        cls.number += 1
+        return self
+
+    def __init__(self, include, late=True, initial=False):
+        if not include:
+            self.text = u""
+        elif include[0] == '<' and include[-1] == '>':
+            self.text = u'#include {0}\n'.format(include)
+            late = False  # system include is never late
+        else:
+            self.text = u'#include "{0}"\n'.format(include)
+        if initial:
+            self.location = self.INITIAL
+        elif late:
+            self.location = self.LATE
+        else:
+            self.location = self.EARLY
+
+    def append(self, text):
+        self.text += text
+
+    def dict_update(self, d, key):
+        """
+        Insert `self` in dict `d` with key `key`. If an element with
+        that key already exists, update only the `location` attribute
+        where the smallest number takes precedence. Return `d[key]`.
+        """
+        if key in d:
+            other = d[key]
+            other.location = min(self.location, other.location)
+            return other
+        else:
+            d[key] = self
+            return self
+
+    def sortkey(self):
+        return self.order
+
+    def __str__(self):
+        text = self.text
+        if not isinstance(text, str):
+            # Python 2: encode unicode -> str
+            text = text.encode('utf-8')
```

Comment:
Shouldn't happen here. The code writer takes Unicode strings and does the encoding on the way out.